client.go

```diff
@@ -43,30 +43,111 @@ func (f backendFactory) New(remote *config.Backend) proxy.Proxy {
 	return f.bf(remote)
 }
 
-func (backendFactory) newChannel(path string) (*amqp.Channel, closer, error) {
+// connectionHandler handles the connection to the amqp backend
+// the connection is shared by all the requests to that backend
+type connectionHandler struct {
+	logger       logging.Logger
+	logPrefix    string
+	mu           *sync.Mutex
+	reconnecting *atomic.Bool
+	conn         connection
+	retries      int
+	backoff      backoff.TimeToWaitBeforeRetry
+}
+
+// newConnectionHandler returns a configured connectionHandler
+func newConnectionHandler(ctx context.Context, l logging.Logger, maxRetries int, strategy, logPrefix string) connectionHandler {
+	c := connectionHandler{
+		logger:       l,
+		logPrefix:    logPrefix,
+		mu:           new(sync.Mutex),
+		reconnecting: new(atomic.Bool),
+		conn:         connection{},
+		retries:      maxRetries,
+		backoff:      backoff.GetByName(strategy),
+	}
+	go func() {
+		<-ctx.Done()
+		c.conn.Close()
+	}()
+	return c
+}
+
+// newConnection should only be used via connect() beacause it doesn't Lock
+// to replace the connection
+func (h *connectionHandler) newConnection(path string) error {
+	c := connection{}
 	conn, err := amqp.Dial(path)
 	if err != nil {
-		return nil, nopCloser, err
+		return err
 	}
+	c.conn = conn
 	ch, err := conn.Channel()
 	if err != nil {
-		return nil, nopCloser, err
+		return err
+	}
+	c.ch = ch
+	h.conn = c
+	return nil
+}
+
+// connect tries to connect to the service with retries given the configuration
+// strategy
+func (h *connectionHandler) connect(dns string) error {
+	var res error
+	h.mu.Lock()
+	defer func() {
+		h.reconnecting.Store(false)
+		h.mu.Unlock()
+	}()
+	h.logger.Debug(h.logPrefix, "reconnecting to host:", dns)
```

Comment:
we think we can remove the `h.mu` and use only the `h.reconnecting` atomic flag. 

Before the defer function we can do:

```go
if !h.reconnecting.CompareAndSwap(false, true) {
    return;
}
```
and then just remove the `h.mu` in other place.

This is a "very tight coupled" concurrent control, and both in `consumer.go`'s `newConsumer` and in `producer.go`'s `newProducer` the very first call that should be done is to the 'h.connect', otherwise, if an error happened before the connect, we could end with `h.reconnecting` not updated to `false` anymore.

I would add a couple of wanning comments inside those both functions.

Reply:
Will leave it as it is now because we don't want to launch too many goroutines unnecessarily. In this case, the tight coupling is for performance reasons.